### PR TITLE
Fix #6319: Add support for `--t: es5` and  `--m es6`

### LINKED ISSUE #6319: Support 'target':'es5' with 'module':'es6' to allow dead code elimination
The import/export statements used in ES6 modules allow tools such as Webpack and Rollup.js to eliminate all the code from unused exports. (For example, see Dr. Axel Rauschmayer's [post](http://www.2ality.com/2015/12/webpack-tree-shaking.html) on this topic)

However, targeting ES5 remove the import/export statements, so these tools cannot remove unused exports.

TypeScript 1.7 included 'es6' and 'es2015' as valid options for the 'module' setting, but the following config:
```javascript
// tsconfig.json
{
  "compilerOptions": {
    "module": "es6",
    "target": "es5"
  }
}
```

Produces this error:
```
error TS1204: Cannot compile modules into 'es6' when targeting 'ES5' or lower.
```
So, it looks like it is not currently possible to generate ES5 code while retaining the import/export statements of ES6 modules.

This is related to issue #4692 that proposes enabling more granular targeting beyond 'ES5' or 'ES6', but in this particular case the two required options already exist in tsconfig.json ('target' and 'module'), but the particular combination of values ('target':'es5', 'module':'es6') doesn't work.

### PR DESCRIPTION
**Pull request by mhegazy**
Fixes #6319.

### CODE REVIEW

File: tests/baselines/reference/es6modulekindWithES5Target9.symbols [hunk 1]
```diff
@@ -0,0 +1,12 @@
+=== tests/cases/compiler/es6modulekindWithES5Target9.ts ===
```
Comment (sheetalkamat) [inline tests/baselines/reference/es6modulekindWithES5Target9.symbols]: Code here looks different from tests/baselines/reference/es6modulekindWithES5Target9.js ?

Comment (mhegazy) [inline tests/baselines/reference/es6modulekindWithES5Target9.symbols]: thanks. earlier files that i have used and did not delete.

File: tests/cases/compiler/es6modulekindWithES5Target.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+declare function foo(...args: any[]): any;
+@foo
+export class D {
+    static s = 0;
+    p = 1;
+    method() { }
+}
+
+class E { }
+export {E};
```
Comment (sheetalkamat) [inline tests/cases/compiler/es6modulekindWithES5Target.ts]: Can you add usage of C and D to  verify exports.C and exports.D is not being used

File: tests/cases/compiler/es6modulekindWithES5Target9.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// @target: es5
+// @module: es2015
+
+import d from "mod";
+
+import {a} from "mod";
+
+import * as M from "mod";
+
```
Comment (sheetalkamat) [inline tests/cases/compiler/es6modulekindWithES5Target9.ts]: Add usage of d, a, M etc to verify exports.identifier is not being emitted

File: tests/cases/compiler/es6modulekindWithES5Target.ts (hunk 2 above)
Comment (mhegazy) [inline tests/cases/compiler/es6modulekindWithES5Target.ts]: sure.

File: tests/cases/compiler/es6modulekindWithES5Target9.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import * as M from "mod";
+
+export {a};
+
+export * from "mod";
+
+export {b} from "mod"
+
+export default d;
```
Comment (sheetalkamat) [inline tests/cases/compiler/es6modulekindWithES5Target9.ts]: Another class decorator test case 
```
                        //  TypeScript                      | Javascript
                        //  --------------------------------|------------------------------------
                        //  @dec                            | let C_1 = class C {
                        //  export default class C {        |   static x() { return C_1.y; }
                        //    static x() { return C.y; }    | };
                        //    static y = 1;                 | let C = C_1;
                        //  }                               | C.y = 1;
                        //                                  | C = C_1 = __decorate([dec], C);
                        //                                  | export default C;
                        //  ---------------------------------------------------------------------
```

File: tests/cases/compiler/es6modulekindWithES5Target9.ts (hunk 3 above)
Comment (mhegazy) [inline tests/cases/compiler/es6modulekindWithES5Target9.ts]: done

File: tests/cases/compiler/es6modulekindWithES5Target9.ts (hunk 4 above)
Comment (mhegazy) [inline tests/cases/compiler/es6modulekindWithES5Target9.ts]: done.

**Comment by mhegazy** [comment]
@sheetalkamat any more comments?

**Comment by sheetalkamat** [comment]
👍

**Comment by AppShipIt** [comment]
CC @Victorystick!

**Comment by Victorystick** [comment]
Woop! This is fantastic news! 🎉 In which stable version can we expect to see these changes?

**Comment by mariusschulz** [comment]
Yey, great to see this feature! 🎉 

@Victorystick The [roadmap](https://github.com/microsoft/typescript/wiki/roadmap) lists it underneath 2.0.

**Comment by AppShipIt** [comment]
It was merged into `master` so I assume 1.9 though and testable in nightly `1.9.0-dev.20160610-1.0`?

**Comment by mhegazy** [comment]
correct. today's build should have the fix. it should be shipped as part of the next release (TS 2.0).

**Pull request merged**